### PR TITLE
[boost] Update to 1.74.0

### PR DESCRIPTION
**Pull request by niclar**
Update to boost-1.74.0

Fixes issue #13033

Tested
x64-windows-static
x64-linux-static

### CODE REVIEW

**Comment by PhoebeHui** [comment]
Thank you @niclar and all for your effort!

Here is CI testing failures, let me know if you need full failure logs.

gamedev-framework:x64-windows-static/gamedev-framework:x64-windows/gamedev-framework:x86-windows:

Failures:
```
D:\installed\x64-windows\include\boost/heap/detail/heap_node.hpp(145): error C2300: 'boost::heap::detail::parent_pointing_heap_node<gf::v1::`anonymous-namespace'::AStarHeapData>': class does not have a destructor called '~Node'
```

bond:arm64-windows/bond:x86-windows:

Failures:
```
D:\buildtrees\bond\src\9.0.2-8585f9f7d7.clean\cpp\inc\bond/core/exception.h(12): fatal error C1083: Cannot open include file: 'boost/locale/encoding_utf.hpp': No such file or directory
```
boost-nowide:x64-osx:

Failures:
```
clang-darwin.compile.c++ /Volumes/data/buildtrees/boost-nowide/x64-osx-dbg/boost/build/e09119a5fb93bc4a6ae7c27175ce30d3/filebuf.o
../src/filebuf.cpp:46:9: error: unknown type name 'constexpr'
        constexpr bool is_in_range(U value)
        ^
../src/filebuf.cpp:62:19: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
            const auto pos = BOOST_NOWIDE_FTELL(file);
                  ^
../src/filebuf.cpp:74:17: error: no matching function for call to 'is_in_range'
            if(!is_in_range<BOOST_NOWIDE_OFF_T>(offset))
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/filebuf.cpp:57:20: error: no matching function for call to 'is_in_range'
            return is_in_range<T>(value) ? static_cast<T>(value) : T(-1);
                   ^~~~~~~~~~~~~~
../src/filebuf.cpp:67:20: note: in instantiation of function template specialization 'boost::nowide::detail::cast_if_valid_or_minus_one<long long, long long>' requested here
            return cast_if_valid_or_minus_one<std::streamoff>(pos);
                   ^
1 warning and 3 errors generated.
```

File: ports/boost-build/CONTROL [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Source: boost-build
-Version: 1.73.0
-Port-Version: 4
+Version: 1.74.0
+Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-build/CONTROL:3]:
```suggestion
```

File: ports/boost-context/CONTROL [hunk 2]
```diff
@@ -1,8 +1,8 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-context
-Version: 1.73.0
-Port-Version: 2
-Build-Depends: boost-assert, boost-build, boost-config, boost-detail, boost-integer, boost-modular-build-helper, boost-pool, boost-predef, boost-smart-ptr, boost-thread, boost-vcpkg-helpers
+Version: 1.74.0
+Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-context/CONTROL:4]:
```suggestion
```

File: ports/boost-coroutine/CONTROL [hunk 3]
```diff
@@ -1,8 +1,8 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-coroutine
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-coroutine/CONTROL:4]:
```suggestion
```

File: ports/boost-fiber/CONTROL [hunk 4]
```diff
@@ -1,8 +1,8 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-fiber
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-fiber/CONTROL:4]:
```suggestion
```

File: ports/boost-filesystem/CONTROL [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-filesystem
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-filesystem/CONTROL:4]:
```suggestion
```

File: ports/boost-iostreams/CONTROL [hunk 6]
```diff
@@ -1,8 +1,8 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-iostreams
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-iostreams/CONTROL:4]:
```suggestion
```

File: ports/boost-locale/CONTROL [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-locale
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-locale/CONTROL:4]:
```suggestion
```

File: ports/boost-log/CONTROL [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-log
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-log/CONTROL:4]:
```suggestion
```

File: ports/boost-modular-build-helper/CONTROL [hunk 9]
```diff
@@ -1,4 +1,4 @@
 Source: boost-modular-build-helper
-Version: 1.73.0
-Port-Version: 6
+Version: 1.74.0
+Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-modular-build-helper/CONTROL:3]:
```suggestion
```

File: ports/boost-mpi/CONTROL [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-mpi
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-mpi/CONTROL:4]:
```suggestion
```

File: ports/boost-parameter-python/CONTROL [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-parameter-python
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-parameter-python/CONTROL:4]:
```suggestion
```

File: ports/boost-poly-collection/CONTROL [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-poly-collection
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-poly-collection/CONTROL:4]:
```suggestion
```

File: ports/boost-python/CONTROL [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-python
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-python/CONTROL:4]:
```suggestion
```

File: ports/boost-stacktrace/CONTROL [hunk 14]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-stacktrace
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-stacktrace/CONTROL:4]:
```suggestion
```

File: ports/boost-test/CONTROL [hunk 15]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-test
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-test/CONTROL:4]:
```suggestion
```

File: ports/boost-wave/CONTROL [hunk 16]
```diff
@@ -1,6 +1,6 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost-wave
-Version: 1.73.0
+Version: 1.74.0
 Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost-wave/CONTROL:4]:
```suggestion
```

File: ports/boost/CONTROL [hunk 17]
```diff
@@ -1,11 +1,10 @@
 # Automatically generated by boost-vcpkg-helpers/generate-ports.ps1
 Source: boost
-Version: 1.73.0
-Port-Version: 3
+Version: 1.74.0
+Port-Version: 1
```
Comment (PhoebeHui) [inline ports/boost/CONTROL:4]:
```suggestion
```

**Review by PhoebeHui — CHANGES_REQUESTED** [review]
Could you remove the Port-Version in affect CONTROL files?

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

File: scripts/boost/post-source-stubs/nowide.cmake [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+
+string(REPLACE "import ../../config/checks/config" "import config/checks/config" _contents "${_contents}")
+
+string(REPLACE "check-target-builds cxx11_moveable_fstreams" "check-target-builds ../check_movable_fstreams.cpp" _contents "${_contents}")
+string(REPLACE "check-target-builds lfs_support" "check-target-builds ../check_lfs_support.cpp" _contents "${_contents}")
+
+file(WRITE "${SOURCE_PATH}/build/Jamfile.v2" "${_contents}")
+file(COPY "${CURRENT_INSTALLED_DIR}/share/boost-config/checks" DESTINATION "${SOURCE_PATH}/build/config")
+
+file(COPY "${SOURCE_PATH}/test/check_lfs_support.cpp" "${SOURCE_PATH}/test/check_movable_fstreams.cpp" DESTINATION "${SOURCE_PATH}/build/config")
```
[the remark this answers is not in the document]
Comment (niclar) [reply scripts/boost/post-source-stubs/nowide.cmake:11]: @PhoebeHui, yes it does

**Comment by niclar** [comment]
@Victor-Bergman can you take a look at ports\0001-Fix-boost-ICU-support.patch & ports\boost-locale\0001-Fix-boost-ICU-support.patch, seems like it prevents x86 deployment of locale

**Comment by niclar** [comment]
-can someone with a osx box take a look at the build ?

**Comment by niclar** [comment]
@jube & @ahugeat can you provide a boost 1.74 patch to gamedev-framework ?

**Comment by ahugeat** [comment]
Hello @niclar,

I have cloned your repo and the compilation of `x64-windows` is working. 
Since we have recently updated gamedev-framework, I have tried with the new version. The triplets `x64-windows`,  `x86-windows`, `x64-windows-static` and  `x86-windows-static` are working. 

Could you try again to compile the gamedev-framework and if it fails, could you send us the complete logs?

**Comment by PhoebeHui** [comment]
Sorry for my late reply.

The failures on CI testing:

gamedev-framework:x64-windows and gamedev-framework:x86-windows:
```
D:\installed\x64-windows\include\boost/heap/detail/heap_node.hpp(145): error C2300: 'boost::heap::detail::parent_pointing_heap_node<gf::v1::`anonymous-namespace'::AStarHeapData>': class does not have a destructor called '~Node'
```
[gamedev-framework.zip](https://github.com/microsoft/vcpkg/files/5450515/gamedev-framework.zip)

bond:x86-windows and bond:arm64-windows:
```
D:\buildtrees\bond\src\9.0.2-8585f9f7d7.clean\cpp\inc\bond/core/exception.h(12): fatal error C1083: Cannot open include file: 'boost/locale/encoding_utf.hpp': No such file or directory
```
[bond.zip](https://github.com/microsoft/vcpkg/files/5450526/bond.zip)

x64-linux and x64-osx:
boost-atomic
boost-container
boost-exception
boost-regex
boost-signals
boost-stacktrace
boost-system

The above port failed with same error:
```
CMake Error at CMakeLists.txt:117:
  Parse error.  Function missing ending ")".  End of file reached.
```

**Comment by longnguyen2004** [comment]
@niclar Seems like you forget a closing paren :)
I'll try to find some time to test gamedev with MSVC and llvm-mingw later

**Comment by victorburckel** [comment]
Hi, don't know if this can help, but it does not seem to be a bug related to the gamedev port. Rather it seems to be a bug with msvc compiling boost binomial heap 1.74. I reported it to see wether it is a bug or not (https://developercommunity.visualstudio.com/content/problem/1240121/problem-compiling-boost-binomial-heap.html). Meanwhile it seems it can be bypassed by patching the following line: https://github.com/boostorg/heap/blob/c0a8773051cd8bc0691ea3c067d59b4d4c5e96dc/include/boost/heap/detail/heap_node.hpp#L145 with a function call ```[](Node* n){ n->~Node(); }(n);```

**Comment by victorburckel** [comment]
For the bond port failure, it seems that adding boost-locale as a dependency (I guess some dependency of boost-config, boost-utility, boost-assign used to depend on that lib before boost 1.74) in the bond CONTROL file seems to fix the issue.

**Comment by victorburckel** [comment]
And for the missing paren, don't know if you already spotted it but it's in your commit fcbee3a you forgot to close the paren
```set(LDFLAGS "${LDFLAGS} <linkflags>-stdlib=libc++"```

**Comment by victorburckel** [comment]
Hello @niclar,

I tried to implement some fixes in #14337, the checks now run ok, can you check if this is ok for you, and maybe integrate them ?

Thanks

**Comment by longnguyen2004** [comment]
If there are no updates from niclar in the coming days then we might as well merge #14337 instead
@victorburckel Can you try submitting a PR to niclar's branch?

**Comment by victorburckel** [comment]
Hi, @longnguyen2004 I'm not able to do a pull request on his repository.

**Comment by niclar** [comment]
@victorburckel , great! -I'll take a look tomorrow..

**Comment by victorburckel** [comment]
@niclar Ok thanks, I took the liberty to push the commits to your repo so that we have the status (should be the same, hope you don't mind).
You'll see there's not much: 
- a patch for boost heap which comes directely from boost develop branch
- the compilation of boost nowide with std=c++11 for apple (the CXXFLAGS in your fcbee3a commit are not taken into account as apple compilation uses clang and there is no such configuration for apple which forwards these flags as for gcc https://github.com/microsoft/vcpkg/blob/00d190a039c78cad4aadfc9a9f3ce8b53165ae6b/ports/boost-modular-build-helper/user-config.jam#L13-L22)
- a new dependency to boost locale for the bond port

**Comment by longnguyen2004** [comment]
@victorburckel @niclar Please also backport [this commit](https://github.com/boostorg/iostreams/pull/106/commits/4e76f73826fd0a7067b837e4850a9051436f5ec5), which fixes a bug when building with llvm-mingw

**Comment by victorburckel** [comment]
@longnguyen2004 Sure, done

**Comment by niclar** [comment]
@victorburckel, it looks good now, thanks. Let's await the test suite run..

**Comment by victorburckel** [comment]
Now there's a link error in rtabmap, but I see the exact same error in other PR, so there are chances the regression does not come from this PR
Here's a sample of the error:
> octomap.lib(OcTreeNode.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in Rtabmap.cpp.obj

**Comment by niclar** [comment]
@seanyen, rtamap debug seem to link to the octomap release library (for x86-windows & x64-windows debug builds), -can you take a look ?

**Comment by victorburckel** [comment]
That's strange, in the output of x64_windows I don't see any errors. I tried to reproduce the error on my machine for the x86 build, but I don't get any error.

**Comment by niclar** [comment]
@victorburckel if you install the octomap lib first (which is apparently done on the CI machine) rtabmap will optionally use it & you'll be able to reproduce the issue

**Comment by niclar** [comment]
@ras0219-msft et.al I can't understand why master is allowed to fail (or isn't staged) ..  so I'll jump out of this rabbit hole until it passes.

**Comment by PhoebeHui** [comment]
I will take a look at the rtabmap failure in CI testing, please ignore it.

**Comment by victorburckel** [comment]
Nice, thanks a lot

**Comment by niclar** [comment]
@PhoebeHui thanks

**Comment by PhoebeHui** [comment]
Feature 'icu' has been tested for boost-locale and boost-regex, all passed with x64-windows.

@ras0219-msft, @BillyONeal, could you help further review?

**Comment by BillyONeal** [comment]
Thanks for your (clearly extensive) help here! 👍

**Pull request merged**